test/Interop/Cxx/namespace/Inputs/namespaces.h

```diff
@@ -0,0 +1,126 @@
+} // namespace BasicChild
+} // namespace BasicParent
+
+namespace BasicParent {
+namespace BasicChild {
+namespace SecondLevelChild {
+inline void inSubNamespaceInRedecl() {}
+} // namespace SecondLevelChild
+} // namespace BasicChild
+} // namespace BasicParent
+
+namespace UniqueOuterName1 {
+namespace SameInnerName {
+inline void basicTestFunction() {}
+} // namespace SameInnerName
+} // namespace UniqueOuterName1
+
+namespace UniqueOuterName2 {
+namespace SameInnerName {
+inline void basicTestFunction() {}
+} // namespace SameInnerName
+} // namespace UniqueOuterName2
+
+namespace HasStructs {
+struct BasicStruct {
+  inline void basicTestFunction() {}
+};
+struct ForwardDeclaredStruct;
+} // namespace HasStructs
+
+struct HasStructs::ForwardDeclaredStruct {};
+
+namespace ParentWithChildWithStructs {
+namespace ChildWithStructs {
+struct ForwardDeclaredChildStruct;
+struct DefinedInDefsStruct;
+
+template <class> struct ForwardDeclaredChildTemplate;
+} // namespace ChildWithStructs
+} // namespace ParentWithChildWithStructs
+
+namespace ParentWithChildWithStructs {
+struct ChildWithStructs::ForwardDeclaredChildStruct {
+  inline void basicTestFunction() {}
+};
+template <class> struct ChildWithStructs::ForwardDeclaredChildTemplate {};
+} // namespace ParentWithChildWithStructs
+
+using NestedForwardDeclaredClassTemplate =
+    ParentWithChildWithStructs::ChildWithStructs::ForwardDeclaredChildTemplate<
+        char>;
+
+namespace ParentWithChildWithFunctions {
+namespace ChildWithFunctions {
+void forwardDeclared();
+void definedInDefs();
+inline void basicTestFunction() {}
+} // namespace ChildWithFunctions
+} // namespace ParentWithChildWithFunctions
+
+namespace ParentWithChildWithFunctions {
+inline void ChildWithFunctions::forwardDeclared(){};
+} // namespace ParentWithChildWithFunctions
+
+namespace ClassTemplates {
+namespace FirstChild {
+struct ForwardDeclared;
+} // namespace FirstChild
+} // namespace ClassTemplates
+
+namespace ClassTemplates {
+namespace SecondChild {
+template <class> struct ClassTemplate {};
+} // namespace SecondChild
+} // namespace ClassTemplates
+
+namespace ClassTemplates {
+namespace FirstChild {
+typedef SecondChild::ClassTemplate<ForwardDeclared> Type;
+inline void takesSomeType(Type) {}
+} // namespace FirstChild
+} // namespace ClassTemplates
+
+namespace ClassTemplates {
+template <class> struct HasSpecialization {};
+template <> struct HasSpecialization<int> {};
+} // namespace ClassTemplates
+
+namespace ClassTemplates {
+using UseTemplate = HasSpecialization<char>;
+// TODO: update tests once these are supported.
+using UseSpecialized = HasSpecialization<int>;
+} // namespace ClassTemplates
+
+namespace FirstWithSameName {
+inline int hasSameName() { return 1; }
+} // namespace FirstWithSameName
+
+namespace SecondWithSameName {
+inline int hasSameName() { return 2; }
+} // namespace SecondWithSameName
+
+#endif // TEST_INTEROP_CXX_NAMESPACE_INPUTS_NAMESPACE_H
```

Comment:
@gribozavr you were correct on both accounts: this was a big overhaul of the tests and it was well worth it. 

I ran into several problems while adding these tests, luckily none were related to namespaces themselves, per se. Anyway, I'm going to try to remove the reliance on #34993 to make the diff easier to read. 

Let me know what you think. I really hope the tests are easier to read now :)

Reply:
(You probably know this already but just in case...) To make the diff easier to read, you can click on the commit where all this is implemented, then the other two commits (which are both separate PRs) won't distract from what's actually being implemented here.